### ISSUE TITLE
gff_filter_by_attribute.py fails badly on GFF3

### ISSUE THREAD

**Issue by peterjc** [galaxyproject/galaxy#3068]
Example test run directly at the command line,

```
$ python tools/filters/gff/gff_filter_by_attribute.py test-data/5.gff3  /dev/stdout "Note__eq__'Ig-like'" '{"Note": "str"}'
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
list index out of range
Filtering with Note=='Ig-like', 
Possible invalid filter condition "Note=='Ig-like'" or non-existent column referenced. See tool tips, syntax and examples.
Skipped 150 invalid lines starting at line #1: "##gff-version 3"
```

The first is `list index out of range` occurs as this tries to split the attribute key-value entries on space (used in GTF and early GFF files), not equals (used in GFF3)

See also #3065 about the missing test files, and #3067 about wrongly reporting comment lines as invalid.

**Comment by peterjc** [comment]
Replacing the split on space with split on equals "fixes" this for GFF3, but breaks on legacy GFF files. Proposed fix working within the current eval approach (including my fix for #3067):

[_UPDATE_ - needs slight tweak for legacy GFF files, pull request to follow]

```
$ git diff
diff --git a/tools/filters/gff/gff_filter_by_attribute.py b/tools/filters/gff/gff_filter_by_attribute.py
index a731df8..2d8bf8d 100644
--- a/tools/filters/gff/gff_filter_by_attribute.py
+++ b/tools/filters/gff/gff_filter_by_attribute.py
@@ -103,10 +103,7 @@ for i, line in enumerate( open( in_fname ) ):
     total_lines += 1
     line = line.rstrip( '\\r\\n' )
     if not line or line.startswith( '#' ):
-        skipped_lines += 1
-        if not invalid_line:
-            first_invalid_line = i + 1
-            invalid_line = line
+        # Ignore blank lines or comments
         continue
     try:
         # Place attribute values into variables with attribute
@@ -114,14 +111,15 @@ for i, line in enumerate( open( in_fname ) ):
         elems = line.split( '\t' )
         attribute_values = {}
         for name_value_pair in elems[8].split(";"):
-            pair = name_value_pair.strip().split(" ")
-            if pair == '':
+            # Split on first equals (GFF3) or space (legacy)
+            i = name_value_pair.replace(" ", "=").find("=")
+            if i == -1:
                 continue
-            name = pair[0].strip()
+            name = name_value_pair[:i].strip()
             if name == '':
                 continue
             # Need to strip double quote from value and typecast.
-            attribute_values[name] = pair[1].strip(" \\"")
+            attribute_values[name] = name_value_pair[i+1:].strip(" \\"")
         %s
         if %s:
             lines_kept += 1
```

Output:

```
$ python tools/filters/gff/gff_filter_by_attribute.py test-data/5.gff  /dev/stdout "score__eq__0" '{"score": "float"}'
chr1    bed2gff CCDS1000.1_cds_0_0_chr1_148325916_f 148325916   148325975   .   +   .   score "0";
chr2    bed2gff CCDS2120.1_cds_0_0_chr2_118288584_f 118288584   118288668   .   +   .   score "0";
chr5    bed2gff CCDS3972.1_cds_0_0_chr5_56241230_f  56241230    56241327    .   +   .   score "0";
chr6    bed2gff CCDS4855.1_cds_0_0_chr6_41411593_f  41411593    41411644    .   +   .   score "0";
chr7    bed2gff CCDS5401.1_cds_0_0_chr7_26907299_r  26907299    26907654    .   +   .   score "0";
chr8    bed2gff CCDS6324.1_cds_0_0_chr8_118881132_r 118881132   118881317   .   +   .   score "0";
chr9    bed2gff CCDS6914.1_cds_0_0_chr9_128764157_f 128764157   128764189   .   +   .   score "0";
chr10   bed2gff CCDS7248.1_cds_0_0_chr10_55251624_r 55251624    55253124    .   +   .   score "0";
chr11   bed2gff CCDS7725.1_cds_0_0_chr11_1731309_r  1731309 1731476 .   +   .   score "0";
chr12   bed2gff CCDS8736.1_cds_0_0_chr12_38440095_r 38440095    38440321    .   +   .   score "0";
chr13   bed2gff CCDS9526.1_cds_0_0_chr13_112381695_f    112381695   112381953   .   +   .   score "0";
chr14   bed2gff CCDS9949.1_cds_0_0_chr14_98710241_r 98710241    98712285    .   +   .   score "0";
chr15   bed2gff CCDS10096.1_cds_0_0_chr15_41486873_r    41486873    41487060    .   +   .   score "0";
chr16   bed2gff CCDS10395.1_cds_0_0_chr16_37430_r   37430   37557   .   +   .   score "0";
chr18   bed2gff CCDS11891.1_cds_0_0_chr18_23786115_r    23786115    23786321    .   +   .   score "0";
chr19   bed2gff CCDS12866.1_cds_0_0_chr19_59068596_f    59068596    59069564    .   +   .   score "0";
chr20   bed2gff CCDS13249.1_cds_0_0_chr20_33330414_r    33330414    33330423    .   +   .   score "0";
chr21   bed2gff CCDS13614.1_cds_0_0_chr21_32707033_f    32707033    32707192    .   +   .   score "0";
chr22   bed2gff CCDS13897.1_cds_0_0_chr22_30120224_f    30120224    30120265    .   +   .   score "0";
chrX    bed2gff CCDS14606.1_cds_0_0_chrX_122745048_f    122745048   122745924   .   +   .   score "0";
Filtering with score==0, 
kept 80.00% of 25 lines.
```

```
$ python tools/filters/gff/gff_filter_by_attribute.py test-data/5.gff3  /dev/stdout "Note__eq__'Ig-like'" '{"Note": "str"}'
ctgA    example motif   33325   35791   .   +   .   ID=Motif:m04;Note=Ig-like
Filtering with Note=='Ig-like', 
kept 0.67% of 150 lines.
```

**Comment by peterjc** [comment]
Thanks to @nsoranzo for pointing me at the test files (#3065), I have started on a pull request to solve this and #3067.
